src/fts.c

```diff
@@ -904,6 +913,20 @@ mem1:				saved_errno = errno;
 		}
 		++nitems;
 	}
+
+	/*
+	 * The only way to distinguish between "no more files" and readdir
+	 * error is checking errno when readdir() returns NULL.
+	 */
+	if (errno) {
+		cur->fts_errno = errno;
+		/*
+		 * If we've not read any items yet, treat
+		 * the error as if we can't access the dir.
+		 */
+		cur->fts_info = nitems ? FTS_ERR : FTS_DNR;
+	}
+
```

Comment:
A part of the original seems to be missing later line 965 : if (!nitems) { ...}
where fts_info can be overridden before returning NULL. The if statement should probably be updated to :
if (type == BREAD && cur->fts_info != FTS_DNR && cur->fts_info != FTS_ERR)